File: 0076-minimum-window-substring/0076-minimum-window-substring.py
from collections import defaultdict

class Solution:
    def minWindow(self, s: str, t: str) -> str:
        mp = defaultdict(int)
        mp2 = defaultdict(int)
        
        for c in t:
            mp[c] += 1
        
        l = 0
        r = 0
        matches = len(mp)
        
        ans = s + "1"
        counter = 0
        
        while r < len(s):
            mp2[s[r]] += 1
            
            if mp2[s[r]] == mp[s[r]]:
                counter += 1
            
            if counter == matches:
                while counter == matches:
                    if r - l + 1 < len(ans):
                        ans = s[l:r + 1]
                    
                    if mp[s[l]] == mp2[s[l]]:
                        counter -= 1
                    
                    mp2[s[l]] -= 1
                    l += 1
            
            r += 1
        
        return "" if ans[-1] == '1' else ans

File: 0076-minimum-window-substring/0076-minimum-window-substring.cpp
class Solution {
public:
    string minWindow(string s, string t) {
        unordered_map<char, int> mt;
        for(const char &e: t)
            mt[e]++;
        unordered_map<char, int> ms;
        int l=0, mini=-1, tl=t.length();
        int matches = 0, ind=0;
        for(int r=0; r<s.length(); r++) {
            //acquire
            ms[s[r]]++;
            if(ms[s[r]]<=mt[s[r]])
                matches++;
            if(matches==tl) {
                if(mini==-1 || r-l+1<mini)
                    mini = r-l+1, ind=l;
                //release
                while(l<=r && matches==tl) {
                    if(r-l+1<mini)
                        mini = r-l+1, ind=l;
                    if(ms[s[l]]==mt[s[l]])
                        matches--;
                    ms[s[l++]]--;
                }
            }
        }
        return (mini==-1) ? "" : s.substr(ind, mini);
    }
};
